altamisa/isatab/models_validation.py

```diff
@@ -0,0 +1,289 @@
+    headers.SCAN_NAME: {PT_DATA_COLLECTION},
+}
+
+# Protocol special case annotations restricted by assay technology
+RESTRICTED_TECH_ANNOS = {
+    # microarray
+    headers.ARRAY_DESIGN_REF: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
+    # gel electrophoresis
+    headers.FIRST_DIMENSION: {GEL_ELECTROPHORESIS},
+    headers.SECOND_DIMENSION: {GEL_ELECTROPHORESIS},
+}
+
+# Protocol special case annotations restricted by protocol type
+RESTRICTED_PROTO_ANNOS = {
+    headers.ARRAY_DESIGN_REF: {PT_HYBRIDIZATION, PT_NUCLEIC_ACID_HYBRIDIZATION},
+    headers.FIRST_DIMENSION: {PT_ELECTROPHORESIS},
+    headers.SECOND_DIMENSION: {PT_ELECTROPHORESIS},
+}
+
+
+# Validator classes --------------------------------------------------------------------
+
+
+class MaterialValidator:
+    """Validator for Material nodes"""
+
+    def __init__(self, assay_info: models.AssayInfo = None):
+        self.assay_info = (assay_info,)
+
+    def validate(self, material: models.Material):
+        """Run Material validations"""
+        # Nothing to do yet
+        # TODO: Move all non-interruptive material validations here
+        pass
+
+
+class ProcessValidator:
+    """Validator for Process nodes"""
+
+    def __init__(
+        self, protocols: Dict[str, models.ProtocolInfo], assay_info: models.AssayInfo = None
+    ):
+        self.protocols = protocols
+        self.assay_info = assay_info
+
+    def validate(self, process: models.Process):
+        """Run Process validations"""
+        valid_ref = self._validate_protocol_ref(process)
+        if valid_ref:
+            self._validate_parameter_values(process)
+            self._validate_name_types(process)
+            self._validate_special_case_annotations(process)
+        else:
+            tpl = (
+                "Can't validate parameter values and names for process with "
+                'undeclared protocol "{}" and name type "{}"'
+            )
+            msg = tpl.format(process.protocol_ref, process.name_type)
+            warnings.warn(msg, IsaValidationWarning)
+
+    def _validate_protocol_ref(self, process: models.Process):
+        # Check if protocol is declared in corresponding study
+        if process.protocol_ref == models.TOKEN_UNKNOWN:
+            return False
+        elif process.protocol_ref not in self.protocols:
+            tpl = 'Protocol "{}" not declared in investigation file'
+            msg = tpl.format(process.protocol_ref)
+            raise ParseIsatabException(msg)
+            # return False when switching to warnings
+        else:
+            return True
+
+    def _validate_parameter_values(self, process: models.Process):
+        # Check if parameter value is declared in corresponding protocol
+        if process.protocol_ref != models.TOKEN_UNKNOWN:
+            for pv in process.parameter_values:
+                if pv.name not in self.protocols[process.protocol_ref].parameters:
+                    tpl = (
```

Comment:
Woudl this fit onto one line?

Reply:
Discussed this quickly with Mikko and we both don't understand how you fit this onto one line. Are you referring to the pv and pv.names?

Reply:
I meant the assignment to `tpl`.